### PR TITLE
Parent-or-else Sampler

### PR DESCRIPTION
**Pull request by ziqizh**
Create a ParentOrElse sampler and corresponding test cases described in the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#parentorelse).

### CODE REVIEW

File: sdk/src/trace/samplers/parent_or_else.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+std::string ParentOrElseSampler::GetDescription() const noexcept
+{
+  // use += to concatenate string for speed
+  std::string description = "ParentOrElse{";
+  description += delegate_sampler_->GetDescription();
+  description += "}";
+  return description;
+}
```
Comment (oulin-coder) [inline sdk/src/trace/samplers/parent_or_else.cc:45]: Similar to probability sampler, compute this at construction time then store in a member variable.

File: sdk/test/trace/parent_or_else_sampler_test.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	auto sampling_result = sampler_off.ShouldSample(nullptr, trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::NOT_RECORD, sampling_result.decision);
+
+	// Case 2: Parent exists and SampledFlag is true
+	auto sampling_result2 = sampler_off.ShouldSample(&parent_context_sampled, trace_id, "", span_kind, view);
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result2.decision);
+
+	// Case 3: Parent exists and SampledFlag is false
+
```
Comment (oulin-coder) [inline sdk/test/trace/parent_or_else_sampler_test.cc:38]: nit: remove extra white line

File: sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#include "opentelemetry/sdk/trace/sampler.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+
+// a placeholder class
```
Comment (oulin-coder) [inline sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h:13]: Add some more info on what this is a placeholder for?

File: sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  explicit ParentOrElseSampler(std::shared_ptr<Sampler> delegate_sampler) noexcept;
+  /** The decision either respects the parent span's sampling decision or delegates to
+   * delegateSampler for root spans
+   * @return Returns NOT_RECORD always
+   */
+  SamplingResult ShouldSample(const SpanContext * /*parent_context*/,
+                              trace_api::TraceId /*trace_id*/,
+                              nostd::string_view /*name*/,
+                              trace_api::SpanKind /*span_kind*/,
+                              const trace_api::KeyValueIterable & /*attributes*/) noexcept override;
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h:43]: You should leave the argument names in, as those are passed on to the delegate sampler.

File: sdk/test/trace/parent_or_else_sampler_test.cc [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+using opentelemetry::sdk::trace::AlwaysOffSampler;
+using opentelemetry::sdk::trace::AlwaysOnSampler;
+using opentelemetry::sdk::trace::Decision;
+using opentelemetry::sdk::trace::ParentOrElseSampler;
+
+TEST(ParentOrElseSampler, ShouldSample)
+{
+  // Initialize a sampler with a AlwaysOffSampler as its delegateSampler
+  ParentOrElseSampler sampler_off(std::make_shared<AlwaysOffSampler>());
+  // Initialize a sampler with a AlwaysOnSampler as its delegateSampler
```
Comment (pyohannes) [inline sdk/test/trace/parent_or_else_sampler_test.cc:16]: I think you can leave those comments out, as it's obvious from the code itself.

File: sdk/test/trace/parent_or_else_sampler_test.cc [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  // Case 2: Parent exists and SampledFlag is true
+  auto sampling_result2 =
+      sampler_off.ShouldSample(&parent_context_sampled, trace_id, "", span_kind, view);
+  ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result2.decision);
+
+  // Case 3: Parent exists and SampledFlag is false
+  auto sampling_result3 =
+      sampler_on.ShouldSample(&parent_context_nonsampled, trace_id, "", span_kind, view);
+  ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result2.decision);
```
Comment (pyohannes) [inline sdk/test/trace/parent_or_else_sampler_test.cc:41]: This should test `sampling_result3`.

File: sdk/test/trace/parent_or_else_sampler_test.cc [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  // Set up parameters
+  opentelemetry::trace::TraceId trace_id;
+  opentelemetry::trace::SpanKind span_kind = opentelemetry::trace::SpanKind::kInternal;
+  using M = std::map<std::string, int>;
+  M m1 = {{}};
+  opentelemetry::trace::KeyValueIterableView<M> view{m1};
+  opentelemetry::sdk::trace::Sampler::SpanContext parent_context_sampled(true, true);
+  opentelemetry::sdk::trace::Sampler::SpanContext parent_context_nonsampled(true, false);
+
+  // Case 1: Parent doesn't exist. Return result of delegateSampler()
```
Comment (pyohannes) [inline sdk/test/trace/parent_or_else_sampler_test.cc:28]: Can we do this test with both `sampler_off` and `sampler_on`?

File: sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+
+/**
+ * A placeholder class that stores the states of a span.
+ * Will be replaced by the real SpanContext class once it is implemented.
+ */
+class Sampler::SpanContext
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h:17]: https://github.com/open-telemetry/opentelemetry-cpp/pull/136/files#r451701975

File: sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+                              const trace_api::KeyValueIterable & attributes) noexcept override;
+
+  /**
+   * @return Description MUST be ParentOrElse{delegate_sampler_.getDescription()}
+   */
+  std::string GetDescription() const noexcept override;
+
+private:
+  std::shared_ptr<Sampler> delegate_sampler_;
+  std::string description_;
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h:52]: We could make both members `const`, as there's no way to change them.

File: sdk/src/trace/samplers/parent_or_else.cc [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    // If parent exists:
+    if (parent_context->sampled_flag)
+    {
+      return {Decision::RECORD_AND_SAMPLE, nullptr};
+    }
+    else
+    {
+      return {Decision::NOT_RECORD, nullptr};
+    }
+  }
```
Comment (pyohannes) [inline sdk/src/trace/samplers/parent_or_else.cc:36]: We could flatten this out, to increase readability:
```c
if (parent_context == nullptr)
{
  return delegate_sampler_->ShouldSample(parent_context, trace_id, name, span_kind, attributes);
}

if (parent_context->sampled_flag)
{
  return {Decision::RECORD_AND_SAMPLE, nullptr};
}

return {Decision::NOT_RECORD, nullptr};
```

**Comment by pyohannes** [comment]
@reyang this should be ready to merge.

File: sdk/src/trace/samplers/parent_or_else.cc [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  return {Decision::NOT_RECORD, nullptr};
+}
+
+std::string ParentOrElseSampler::GetDescription() const noexcept
+{
+  return description_;
+}
+}  // namespace trace
+}  // namespace sdk
+OPENTELEMETRY_END_NAMESPACE
```
Comment (reyang) [inline sdk/src/trace/samplers/parent_or_else.cc:41]: nit: would you add an empty line before EOF?

File: sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+{
+public:
+  inline explicit SpanContext(bool is_recording, bool sampled_flag)
+      : is_recording(is_recording), sampled_flag(sampled_flag)
+  {}
+
+  bool is_recording;
+  bool sampled_flag;
+};
+/**
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h:27]: nit: probably add an empty line here.

Comment (ziqizh) [reply sdk/include/opentelemetry/sdk/trace/samplers/parent_or_else.h:27]: Done.

**Comment by reyang** [comment]
Please rebase after fixing the nits.

**Pull request merged**